### ISSUE TITLE
Local dependencies broken in npm v3

### ISSUE THREAD

**Issue by timoxley** [npm/npm#9205]
npm 3 crashes with an `ENOENT` error during an `npm install` if your app contains relative `file:` path local dependencies that have the dependency structure outlined below.

I believe this is caused by npm 3 looking for the child dependency relative to location of the parent inside node_modules, i.e. '../child' == 'node_modules/parent/../child' rather than relative to the location of the local file's source.

Tested on npm v3.2.2.
The structure installs just fine on npm v2.13.5.
I have cleared the cache and *not* cleared the cache between tests and witnessed no difference in result.

```dot
digraph {
  root -> A
  root -> B
  A -> B
  A -> C
  C -> B
}
```
![image](https://cloud.githubusercontent.com/assets/43438/9153410/77e3caaa-3e19-11e5-84ae-4fd204e0e201.png)

### Note that almost any other configuration will work. 

For example:

It works fine if you remove the root package's dependency on `B` 

```dot
digraph {
  root -> A
  // root -> B
  A -> B
  A -> C
  C -> B
}
```

It works if you flip the dependency between `B` and `C` i.e. `B -> C` instead of `C -> B`:

```dot
digraph {
  root -> A
  root -> B
  A -> B
  A -> C
  // C -> B
  B -> C
}
```

It works if you completely remove the dependency from `C -> B`:

```dot
digraph {
  root -> A
  root -> B
  A -> B
  A -> C
  // C -> B
}
```

----

### Concrete Example

See [timoxley/npm3-localdeps-issue](https://github.com/timoxley/npm3-localdeps-issue) for a concrete example of the failing structure.

```
git clone git@github.com:timoxley/npm3-localdeps-issue.git
cd npm3-localdeps-issue
npm install
```

Should produce the following error:
```
npm install
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm3-localdeps/node_modules/banana
npm ERR! Darwin 14.4.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v2.0.1
npm ERR! npm  v3.2.2
npm ERR! path /Users/timoxley/Projects/tests/npm3-localdeps/node_modules/banana
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open
...
```

npm 3 also doesn't want to install this structure from the registry, with a similar yet different error:

`npm install npm3-localdeps-issue`

```
> npm install  npm3-localdeps-issue
npm http request GET https://registry.npmjs.org/npm3-localdeps-issue
npm http 200 https://registry.npmjs.org/npm3-localdeps-issue
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm3-localdeps-changes/node_modules/npm3-localdeps-issue/apple
npm WARN install Couldn't install optional dependency: ENOENT: no such file or directory, open '/Users/timoxley/Projects/tests/npm3-localdeps-changes/node_modules/npm3-localdeps-issue/apple'
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm3-localdeps-changes/node_modules/npm3-localdeps-issue/banana
npm WARN install Couldn't install optional dependency: ENOENT: no such file or directory, open '/Users/timoxley/Projects/tests/npm3-localdeps-changes/node_modules/npm3-localdeps-issue/banana'
npm WARN EPACKAGEJSON npm3-localdeps-changes@1.0.0 No description
npm WARN EPACKAGEJSON npm3-localdeps-changes@1.0.0 No repository field.
```

Also note that npm 3 incorrectly reports local dependencies to be optional dependencies:

```
 npm WARN install Couldn't install optional dependency: ENOENT: no such file or directory,
```

To be fair though, despite functioning just fine locally, npm 2 also feels bummed about installing local deps from the registry:

```
> npm-next install npm3-localdeps-issue
npm WARN package.json npm3-localdeps-changes@1.0.0 No description
npm WARN package.json npm3-localdeps-changes@1.0.0 No repository field.
npm WARN package.json npm3-localdeps-changes@1.0.0 No README data
npm http request GET https://registry.npmjs.org/npm3-localdeps-issue
npm http 304 https://registry.npmjs.org/npm3-localdeps-issue
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm3-localdeps-changes/npm3-localdeps-issue@*/apple
npm ERR! addLocal Could not install /Users/timoxley/Projects/tests/npm3-localdeps-changes/npm3-localdeps-issue@*/banana
npm ERR! Darwin 14.4.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/lib/node_modules/npm-next/node_modules/npm/cli.js" "install" "npm3-localdeps-issue"
npm ERR! node v2.0.1
npm ERR! npm  v2.13.5
npm ERR! path /Users/timoxley/Projects/tests/npm3-localdeps-changes/npm3-localdeps-issue@*/apple
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open '/Users/timoxley/Projects/tests/npm3-localdeps-changes/npm3-localdeps-issue@*/apple'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/timoxley/Projects/tests/npm3-localdeps-changes/npm-debug.log
```

**Comment by othiym23** [comment]
There's a piece that's missing from `npm@2` which I tried to get in when @dylang and I landed local dependencies, and which would have bearing here: I don't think that a `package.json` with a local dependency in it should be publishable. _Maaaaybe_ it's OK if the paths are all relative and point to directories that will be included in the package, but that seems like a feature that's more confusing and complicated than it's worth. A much simpler feature is just to error with something like "can't publish with local dependencies" and exit if you try to publish with `file:` specifiers in `package.json`.

I'll leave the rest of this up to @iarna to respond to, although your reasoning seems sound to me. There's still some work to be done to figure out how to make local dependencies and `npm link` complementary instead of largely disjoint features.

**Comment by iarna** [comment]
It's funny that after you added this issue, while constructing a test for an unrelated bug I was then burned by exactly this bug. =D

**Comment by othiym23** [comment]
Fixed via 4468c92eccd0814b567606897e0792e49640e056, included in `npm@3.3.1`. Thanks for the well-illustrated bug report, Tim! Let us know if you notice any further issues with npm 3 and local dependencies!

**Comment by walea3** [comment]
hi @othiym23 @iarna, I'm currently able to reproduce this issue with `@3.3.1` and `@latest` - using @timoxley's example, run `npm install https://github.com/timoxley/npm3-localdeps-issue.git`

However, it's not an issue if you clone the repo and then `npm install ../npm3-localdeps-issue`, for example.

It looks as though the local dependencies aren't being resolved from the temp directory the repo is cloned into but the final directory _(`node_modules/npm3-localdeps-issue`)_ which hasn't been created yet.
